wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+    slots <- vectorOf (length pendingTxs) (choose (lowerBound, lowerBound + 10))
+    retries    <- vectorOf (length pendingTxs) (choose (0, maxRetries))
+    return $ Scheduler $ List.foldl' updateFn mempty (zip3 slots pendingTxs retries)
+    where
+        updateFn acc (slot, (txId, txAux), retries) =
+            let s = Scheduled (txId, txAux, Schedule retries)
+            in IntMap.insertWith mappend slot [s] acc
+
+instance Arbitrary WalletSubmissionState where
+    arbitrary = do
+        pending   <- genPending (Core.ProtocolMagic 0)
+        slot      <- fmap getPositive arbitrary
+        scheduler <- genScheduler pending slot
+        return $ WalletSubmissionState pending scheduler slot
+
+genWalletSubmission :: ResubmissionFunction m -> Gen (WalletSubmission m)
+genWalletSubmission rho =
+    WalletSubmission <$> pure rho <*> arbitrary
+
+
+--
+-- Public API, as written in the spec
+--
+
+newWalletSubmission :: ResubmissionFunction m -> WalletSubmission m
+newWalletSubmission resubmissionFunction = WalletSubmission {
+      _wsResubmissionFunction = resubmissionFunction
+    , _wsState = newEmptyState
+    }
+    where
+        newEmptyState :: WalletSubmissionState
+        newEmptyState = WalletSubmissionState {
+              _wssPendingSet  = emptyPending
+            , _wssCurrentSlot = 0
+            , _wssScheduler   = Scheduler IntMap.empty
+            }
+
+-- | Informs the 'WalletSubmission' layer about new 'Pending' transactions.
+addPending :: WalletSubmission m -> Pending -> WalletSubmission m
+addPending ws newPending =
+    let ws' = ws & over (wsState . wssPendingSet) (unionPending newPending)
+    in schedule ws' newPending
+
+-- | Removes the input 'Pending' from the local 'WalletSubmission' pending set.
+remPending :: WalletSubmission m -> Pending -> WalletSubmission m
+remPending ws updatedPending =
+    ws & over (wsState . wssPendingSet) (flip differencePending updatedPending)
+
+-- | A \"tick\" of the scheduler.
+-- Returns the set transactions which needs to be droppped by the system as
+-- they likely exceeded the submission count and they have no chance to be
+-- adopted in a block, or that have simply been successfully retrasmitted.
+-- @N.B.@ The returned 'WalletSubmission' does not come with an already-pruned
+-- local 'Pending' set. It will be caller's responsibility to call 'remPending'
+-- on the 'WalletSubmission'.
+--
+tick :: Monad m
+     => WalletSubmission m
+     -- ^ The current 'WalletSubmission'.
+     -> m (Evicted, WalletSubmission m)
+     -- ^ The set of transactions upper layers will need to drop, together
+     -- with the new 'WalletSubmission'.
+tick ws = do
+    let wss               = ws  ^. wsState
+        currentSlot       = wss ^. wssCurrentSlot
+        scheduler         = ws  ^. getScheduler
+        rho               = _wsResubmissionFunction ws
+        dueNow            = dueThisSlot currentSlot ws
+    (toPrune, scheduler') <- rho currentSlot dueNow scheduler
```

Comment:
If we change the naming of `Scheduler`, we should remember to change it here also. (And I find it equally confusing here; `scheduler` isn't actually a scheduler, but rather a schedule).